crates/rust-analyzer/src/caps.rs

```diff
@@ -4,16 +4,47 @@ use std::env;
-    CodeLensOptions, CompletionOptions, DocumentOnTypeFormattingOptions,
-    FoldingRangeProviderCapability, ImplementationProviderCapability, RenameOptions,
-    RenameProviderCapability, SaveOptions, SelectionRangeProviderCapability,
-    SemanticTokensDocumentProvider, SemanticTokensLegend, SemanticTokensOptions,
-    ServerCapabilities, SignatureHelpOptions, TextDocumentSyncCapability, TextDocumentSyncKind,
-    TextDocumentSyncOptions, TypeDefinitionProviderCapability, WorkDoneProgressOptions,
+    CallHierarchyServerCapability, ClientCapabilities, CodeActionOptions,
+    CodeActionProviderCapability, CodeLensOptions, CompletionOptions,
+    DocumentOnTypeFormattingOptions, FoldingRangeProviderCapability,
+    ImplementationProviderCapability, RenameOptions, RenameProviderCapability, SaveOptions,
+    SelectionRangeProviderCapability, SemanticTokensDocumentProvider, SemanticTokensLegend,
+    SemanticTokensOptions, ServerCapabilities, SignatureHelpOptions, TextDocumentSyncCapability,
+    TextDocumentSyncKind, TextDocumentSyncOptions, TypeDefinitionProviderCapability,
+    WorkDoneProgressOptions,
 };
 
-pub fn server_capabilities() -> ServerCapabilities {
+pub fn server_capabilities(client_caps: &ClientCapabilities) -> ServerCapabilities {
+    let mut code_action_provider = CodeActionProviderCapability::Simple(true);
+
+    match client_caps.text_document.as_ref() {
+        Some(it) => {
+            match it.code_action.as_ref().and_then(|c| c.code_action_literal_support.as_ref()) {
+                Some(_literal_support) => {
+                    code_action_provider =
+                        CodeActionProviderCapability::Options(CodeActionOptions {
+                            // Advertise support for all built-in CodeActionKinds.
+                            // Ideally we would base this off of the client capabilities
+                            // but the client is supposed to fall back gracefully for unknown values.
+                            code_action_kinds: Some(vec![
+                                lsp_types::code_action_kind::EMPTY.to_string(),
+                                lsp_types::code_action_kind::QUICKFIX.to_string(),
+                                lsp_types::code_action_kind::REFACTOR.to_string(),
+                                lsp_types::code_action_kind::REFACTOR_EXTRACT.to_string(),
+                                lsp_types::code_action_kind::REFACTOR_INLINE.to_string(),
+                                lsp_types::code_action_kind::REFACTOR_REWRITE.to_string(),
+                                lsp_types::code_action_kind::SOURCE.to_string(),
+                                lsp_types::code_action_kind::SOURCE_ORGANIZE_IMPORTS.to_string(),
+                            ]),
+                            work_done_progress_options: Default::default(),
+                        });
+                }
+                None => {}
+            }
+        }
+        None => {}
+    };
```

Comment:
This is ugly but I expect a lot more to go into the top-level (most caps will depend on the presence of `caps.text_document`. Probably most capabilities should be refactored into functions that take a `&TextDocumentClientCapability` or maybe the whole client capabilities and they all repeat the check or something...